sql/coord.cc

```diff
@@ -0,0 +1,385 @@
+stringEqual(const char * other) const
+{
+    return compareStringNullTerminated(other) == 0;
+}
+
+bool
+Coord::
+stringLess(const std::string & other) const
+{
+    return compareString(other.data(), other.size()) < 0;
+}
+
+bool
+Coord::
+stringLess(const Utf8String & other) const
+{
+    return compareString(other.rawData(), other.rawLength()) < 0;
+}
+
+bool
+Coord::
+stringLess(const char * other) const
+{
+    return compareStringNullTerminated(other) < 0;
+}
+
+bool
+Coord::
+stringGreaterEqual(const std::string & other) const
+{
+    return compareString(other.data(), other.size()) >= 0;
+}
+
+bool
+Coord::
+stringGreaterEqual(const Utf8String & other) const
+{
+    return compareString(other.rawData(), other.rawLength()) >= 0;
+}
+
+bool
+Coord::
+stringGreaterEqual(const char * other) const
+{
+    return compareStringNullTerminated(other) >= 0;
+}
+    
+bool
+Coord::
+operator == (const Coord & other) const
+{
+    return dataLength() == other.dataLength()
+        && compareString(other.data(), other.dataLength()) == 0;
+}
+
+bool
+Coord::
+operator != (const Coord & other) const
+{
+    return ! operator == (other);
+}
+
+bool
+Coord::
+operator <  (const Coord & other) const
+{
+    return compareString(other.data(), other.dataLength()) < 0;
+}
+
+Utf8String
+Coord::
+toUtf8String() const
+{
+    if (complex_)
+        return getComplex();
+    else return Utf8String(data(), dataLength(), true /* is valid UTF-8 */);
+}
+
+std::string
+Coord::
+toString() const
+{
+    return toUtf8String().stealRawString();
+}
+
+constexpr HashSeed defaultSeedStable { .i64 = { 0x1958DF94340e7cbaULL, 0x8928Fc8B84a0ULL } };
+
+uint64_t
+Coord::
+hash() const
+{
+    return Id(data(), dataLength()).hash();
+    //return ::mldb_siphash24(str.rawData(), str.rawLength(), defaultSeedStable.b);
+}
+
+bool
+Coord::
+empty() const
+{
+    return complex_ == 0 && simpleLen_ == 0;
+}
+
+Coord
+Coord::
+operator + (const Coord & other) const
+{
+    if (empty())
+        return std::move(other);
+    if (other.empty())
+        return *this;
+    return toUtf8String() + "." + other.toUtf8String();
+}
+
+Coord
+Coord::
+operator + (Coord && other) const
+{
+    if (empty())
+        return std::move(other);
+    if (other.empty())
+        return *this;
+    return toUtf8String() + "." + other.toUtf8String();
+}
+
+Coord::
+operator RowHash() const
+{
+    return RowHash(hash());
+}
+
+Coord::
+operator ColumnHash() const
+{
+    return ColumnHash(hash());
+}
+
+void
+Coord::
+complexDestroy()
+{
+    getComplex().~Utf8String();
+}
+
+void
+Coord::
+complexCopyConstruct(const Coord & other)
+{
+    new (&str.str) Utf8String(other.getComplex());
+}
+
+void
+Coord::
+complexMoveConstruct(Coord && other)
+{
+    new (&str.str) Utf8String(std::move(other.getComplex()));
+}
+
+void
+Coord::
+initString(Utf8String str)
+{
+    ExcAssertEqual(strlen(str.rawData()), str.rawLength());
+    words[0] = words[1] = words[2] = words[3] = 0;
+    if (str.rawLength() <= 31) {
```

Comment:
Yes.  I figured it out when my confusion about bitfields / initialization was cleared out.